src/storage/src/render/multi_worker_persist_sink.rs

```diff
@@ -0,0 +1,1123 @@
+
+        if !active_worker {
+            // The non-active workers report that they are done snapshotting.
+            return;
+        }
+
+        // Initialize this operators's `upper` to the `upper` of the persist shard we are writing
+        // to. Data from the source not beyond this time will be dropped, as it has already
+        // been persisted.
+        // In the future, sources will avoid passing through data not beyond this upper
+        let mut current_upper = {
+            // TODO(aljoscha): We need to figure out what to do with error
+            // results from these calls.
+            let persist_client = persist_clients
+                .open(persist_location)
+                .await
+                .expect("could not open persist client");
+
+            let write = persist_client
+                .open_writer::<SourceData, (), Timestamp, Diff>(
+                    shard_id,
+                    &format!(
+                        "compute::persist_sink::mint_batch_descriptions {}",
+                        collection_id
+                    ),
+                    Arc::new(target_relation_desc),
+                    Arc::new(UnitSchema),
+                )
+                .await
+                .expect("could not open persist shard");
+
+            let upper = write.upper().clone();
+            // explicitly expire the once-used write handle.
+            write.expire().await;
+            upper
+        };
+
+        // The current input frontiers.
+        let mut desired_frontier;
+
+        loop {
+            if let Some(event) = desired_input.next().await {
+                match event {
+                    Event::Data(_cap, _data) => {
+                        // Just read away data.
+                        // TODO(guswynn): this, and the same code in the compute version of this
+                        // code, is inefficient, and can be improved.
+                        // See
+                        // <https://github.com/MaterializeInc/materialize/pull/17589#discussion_r1106016139>
+                        // for more info.
+                        continue;
+                    }
+                    Event::Progress(frontier) => {
+                        desired_frontier = frontier;
+                    }
+                }
+            } else {
+                // Input is exhausted, so we can shut down.
+                return;
+            };
+
+            // If the new frontier for the data input has progressed, produce a batch description.
+            if PartialOrder::less_than(&current_upper, &desired_frontier) {
+                // The maximal description range we can produce.
+                let batch_description = (current_upper.to_owned(), desired_frontier.to_owned());
+
+                let lower = batch_description.0.first().unwrap();
+                let batch_ts = batch_description.0.first().unwrap().clone();
+
+                let cap = cap_set
```

Comment:
NIT: this will probably be much shorter with:

```rust
let Ok(cap) = cap_set.try_delayed(&batch_ts) else {
    panic!(
        "minter cannot delay {cap_set:?} to {lower:?}. Likely because \
        we already emitted a batch description and delayed."
    );
};
```

Reply:
Im anti `let-else`!

Reply:
Spicy take. Still, a match is much sorter than the closures but no strong opinions

```rust
let cap = cap_set.try_delayed(&batch_ts) {
    Ok(cap) => cap,
    Err(_) => panic!(
        "minter cannot delay {cap_set:?} to {lower:?}. Likely because \
        we already emitted a batch description and delayed."
    ),
};
```